zng/docs/spec.md

```diff
@@ -350,121 +470,126 @@ previously defined type, the appropriate typedef control code must
 
 Since N, the byte length of any of these container values, is known,
 there is no need to encode a count of the
 elements present.  Also, since the type ID is implied by the typedef
-of any container type, each value is encoded without its type ID.
+of any complex type, each value is encoded without its type ID.
 
-The concatenation of elements is encoded as a sequence of "tag-counted" values.
-A tag carries both the length information of the corresponding value as well
-a "container bit" to differentiate between primitive values and container values
-without having to refer to the implied type.  This admits an efficient implementation
-for traversing the values, inclusive of recursive traversal of container values,
-whereby the inner loop need not consult and interpret the type ID of each element.
+For sets, the concatenation of elements must be normalized so that the
+sequence of bytes encoding each element's tag-counted value is
+lexicographically greater than that of the preceding element.
 
-The tag encodes the length N of the value and indicates whether
-it is a primitive value or a container value.
-The length is offset by 1 whereby length of 0 represents an unset value
-analogous to null in JSON.
-The container bit is 1 for container values and 0 for primitive values.
-The tag is defined as
-```
-2*(N+1) + the container bit
-```
-and is encoded as a `uvarint`.
+A union value is encoded as a container with two elements. The first
+element is the `uvarint` encoding of the index determining the type of
+the value in reference to the union type, and the second element is
+the value encoded according to that type.
 
-For example, tag value 0 is an unset primitive value and tag value 1
-is an unset container value.  Tag value 2 is a length zero primitive
-value, e.g., it could represent empty string.  Tag value 3 is a length
-zero container, such as an empty array or record with no fields.  Tag
-value 4 is a length 1 primitive value, e.g., it would represent the
-boolean "true" if followed by byte value 1 in the context of type ID 0
-(i.e., the type ID for boolean).
+An enumeration value is simply a single, small integer encoded as length-N
+integer that represents the positional index of enum identifier and value as it
+is defined in the enum typedef.
 
-Following the tag encoding is the value encoded in N bytes as described above.
+A map value is encoded as a container as a sequence of alternating tag-encoded
+key and value. The concatenation of elements must be normalized so that the
+sequence of bytes encoding each tag-counted key (of the key/value pair) is
+lexicographically greater than that of the preceding key (of the preceding
+key/value pair).
 
-For sets, the concatenation of elements must be normalized so that the
-sequence of bytes encoding each element's tag-counted value is
-lexicographically greater than that of the preceding element.
+> XXX it's not clear that we should require sorted keys and give up
```

Comment:
I'm not understanding this sentence... is the deletion of the paragraph describing set normalization suggesting that we remove that requirement ? (I don't see how we can, hence part of my confusion here).

Reply:
This was moved to line 573.  The XXX refers to map values.  Maybe we can discuss at sync-up?

Reply:
Ah hadn't spotted that the paragraph moved. This makes sense.